Hacking

Fortune117 opened the pull request:
# Node Based hacking

### Purpose
Implements hacking as a node based system. Players can 'rewire' objects to change the control flow of them, as well as add hacking devices to act as I/O for the device.
Examples:
https://cdn.discordapp.com/attachments/315652489544335361/702761159027327016/hacking_demo.mp4
https://cdn.discordapp.com/attachments/315652489544335361/703663984880123924/hacking_devices.mp4

This also adds a Radio Manager, which can be used to send radio signals between devices with RadioReceiver and RadioMessager components.

These devices have customizable frequency broadcast ranges, amplitudes, sensitivity, etc. 
There's still more to do with this manager, but that can be done in further, separate pull requests.

As an example to demonstrate the radio system and hacking device system, a Remote Signaller is included in this PR. It can be inserted into the hacking panel of an object and has both an input and output node.

Fixes #3315

### Notes:
This is the earliest working version I wanted to upload. There's lots of things missing and there's likely to be bugs. The code needs cleaning up, and there are some not insignificant changes that need to be made to the UI. But it's working, and Redline wanted me to upload it.

### Please make sure you have followed the self checks below before submitting a PR:

- [ x ] Code is sufficiently commented
- [ x ] Code is indented with tabs and not spaces
- [  x ] The PR does not include any unnecessary .meta, .prefab or **.unity (scene) changes**
- [ x ] The PR does not bring up any new compile errors
- [ x ] The PR has been tested in editor
- [ x ] Any new files are named using PascalCase (to avoid issues on case sensitive file systems)
- [ x ] Any new / changed components follow the [Component Development Checklist](https://github.com/unitystation/unitystation/wiki/Component-Development-Checklist)
- [ x ] Any new objects / items follow the [Creating Items and Objects Guide](https://github.com/unitystation/unitystation/wiki/Creating-Items-and-Objects%3A-Now-With-Prefab-Variants) (especially concerning the use of prefab variants)
- [ x ] The PR has been tested in multiplayer (with 2 clients and late joiners, if applicable)

Fortune117 committed:
Beginning of Hacking

Fortune117 committed:
Visible UI Added

Fortune117 committed:
Fixed NetTab issue, working prototype (server host only)

Fortune117 committed:
Added Networking, needs testing.

Fortune117 committed:
Minor Bugfix

Fortune117 committed:
Merge remote-tracking branch 'upstream/develop' into hacking

Fortune117 committed:
Hacking now syncs client views, also added Radios and a Remote Signaller

Fortune117 committed:
Hacking Devices can now be added to hacking panels. NEEDS MORE WORK.

Fortune117 committed:
Update GUI_HackingWire.cs

Fortune117 committed:
Devices are now networked and stored in the panel.

There will be bugs, I haven't limited the number of devices you can put in a panel (although, it should stop at 4, since that's how many slots it has).

Fortune117 committed:
Added hacking to all door prefabs.

daalta commented on UnityProject/Assets/Scripts/Devices/RemoteSignaller.cs:
```diff
@@ -0,0 +1,100 @@
+[RequireComponent(typeof(RadioMessager))]
+[RequireComponent(typeof(RadioReceiver))]
+[RequireComponent(typeof(HackingDevice))]
+public class RemoteSignaller : NetworkBehaviour, IInteractable<HandActivate>, IServerSpawn
```
Maybe add a short description of what the RemoteSignaler is?

Also... I think it's "Signaler" with a single L, not "Signa**ll**er"

You typed it correctly in the chat message :)

Fortune117 committed:
Cleaned up code, added basic textures, still looks ugly ngl

Removed a bunch of no longer needed debug statements.

Fortune117 committed:
Doors randomised, debug messages removed, prone to breaking, lets do this

Fortune117 committed:
Added randomised door nodes, requires wire cutters, lookin good

RedlineTriad commented on UnityProject/Assets/Scripts/Devices/RemoteSignaller.cs:
```diff
@@ -0,0 +1,100 @@
+	{
+		if (IsOn)
+		{
+			if (interaction.Intent == Intent.Harm)
+			{
+				radioMessager.SendSignal();
+			}
+			else if (interaction.Intent == Intent.Disarm)
+			{
+				SyncIsOn(IsOn, !IsOn);
+			}
+			else if (interaction.Intent == Intent.Help)
+			{
+				//Ad frequency change code here.
+			}
```
Why not use a `switch` statement?

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/HackingDevice.cs:
```diff
@@ -0,0 +1,52 @@
+
+	public void OnSpawnServer(SpawnInfo info)
+	{
+		inputNode = new HackingNode();
+		inputNode.AddToInputMethods(onInputReceived.Invoke);
+		inputNode.HiddenLabel = "Input";
+		inputNode.IsInput = true;
+		inputNode.IsDeviceNode = true;
+		outputNode = new HackingNode();
+		outputNode.HiddenLabel = "Output";
+		outputNode.IsOutput = true;
+		outputNode.IsDeviceNode = true;
```
```suggestion
		inputNode = new HackingNode()
		{
			HiddenLabel = "Input",
			IsInput = true,
			IsDeviceNode = true,
		};
		inputNode.AddToInputMethods(onInputReceived.Invoke);
		outputNode = new HackingNode()
		{
			HiddenLabel = "Output",
			IsOutput = true,
			IsDeviceNode = true,
		};
```
Also maybe create a constructor instead of manually setting values.

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/HackingDevice.cs:
```diff
@@ -0,0 +1,52 @@
+	{
+		if (!CustomNetworkManager.IsServer)
+		{
+			inputNode = new HackingNode();
+			inputNode.HiddenLabel = "Input";
+			inputNode.IsInput = true;
+			inputNode.IsDeviceNode = true;
+			outputNode = new HackingNode();
+			outputNode.HiddenLabel = "Output";
+			outputNode.IsOutput = true;
+			outputNode.IsDeviceNode = true;
```
Same here.

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/HackingProcesses/HackingProcessDoorSimple.cs:
```diff
@@ -0,0 +1,214 @@
+	{
+		get
+		{
+			if (!controller)
```
```suggestion
			if (controller != null)
```
Makes intentions more clear.

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/HackingProcesses/HackingProcessDoorSimple.cs:
```diff
@@ -0,0 +1,214 @@
+	{
+		get
+		{
+			if (!intDoor)
```
```suggestion
			if (intDoor != null)
```

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/HackingProcesses/HackingProcessDoorSimple.cs:
```diff
@@ -0,0 +1,214 @@
+		Shuffle(infList);
+		foreach (HackingNodeInfo inf in infList)
+		{
+				HackingNode newNode = new HackingNode();
+				newNode.IsInput = inf.IsInput;
+				newNode.IsOutput = inf.IsOutput;
+				newNode.IsDeviceNode = inf.IsDeviceNode;
+				newNode.InternalIdentifier = inf.InternalIdentifier;
+				newNode.HiddenLabel = inf.HiddenLabel;
+				newNode.PublicLabel = inf.PublicLabel;
```
Use initializer syntax instead.

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/NodeSystem/HackingNode.cs:
```diff
@@ -0,0 +1,154 @@
+	public string InternalIdentifier { get; set; }
+
+	//Bools for whether the node is an input or output node. I'm not exactly sure how these will work, so maybe eventually there could be a situation where a node is both? Dunno
+	public	bool IsInput { get; set; }
+
+	public bool IsOutput { get; set; }
```
Maybe use an `enum` instead?

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/NodeSystem/HackingNode.cs:
```diff
@@ -0,0 +1,154 @@
+
+	public virtual void AddWireCutCallback(Action method)
+	{
+		Action<GameObject> methodWrapped = delegate (GameObject gameObject)
+		{
+			method();
+		};
```
You don't need to use `delegate`, you should be able to do
```suggestion
		Action<GameObject> methodWrapped = (GameObject gameObject) => method();
```

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/NodeSystem/HackingNode.cs:
```diff
@@ -0,0 +1,154 @@
+
+	public virtual void AddToInputMethods(Action method)
+	{
+		Action<GameObject> methodWrapped = delegate (GameObject gameObject)
+		{
+			method();
+		};
+
```
```suggestion
		Action<GameObject> methodWrapped = (GameObject gameObject) => method();

```

RedlineTriad commented on UnityProject/Assets/Scripts/Hacking/NodeSystem/HackingNode.cs:
```diff
@@ -0,0 +1,154 @@
+
+	public virtual void RemoveFromInputMethods(Action method)
+	{
+		Action<GameObject> methodWrapped = delegate (GameObject gameObject)
+		{
+			method();
+		};
```
```suggestion
		Action<GameObject> methodWrapped = (GameObject gameObject) => method();
```

RedlineTriad commented on UnityProject/Assets/Scripts/Managers/RadioManager.cs:
```diff
@@ -0,0 +1,163 @@
+	{
+		get
+		{
+			if (!radioManager)
```
```suggestion
			if (radioManager != null)
```

RedlineTriad commented on UnityProject/Assets/Scripts/Managers/RadioManager.cs:
```diff
@@ -0,0 +1,163 @@
+	//Degrade the message in a sent signal if it's below the sensitivity of the receiver (up to a certain range).
+	//i.e. add corrupted text and stuff
+
+	private bool CanReceiverReceiverSignal(RadioReceiver rec, RadioSignal signal, RadioMessager originator)
```
This method is way to complex, try to break it up into separate smaller ones.

RedlineTriad commented on UnityProject/Assets/Scripts/Managers/RadioManager.cs:
```diff
@@ -0,0 +1,163 @@
+		}
+	}
+
+	public void SendRadioSignalFromLocation(RadioSignal signal, Vector2 origin)
+	{
+
+	}
```
What is this for?

RedlineTriad commented on UnityProject/Assets/Scripts/Player/PlayerScript.cs:
```diff
@@ -225,6 +225,8 @@ public bool IsDeadOrGhost
 		}
 	}
 
+	public object Chat { get; internal set; }
```
Why use `internal`?

RedlineTriad commented on UnityProject/Assets/Scripts/Radio/RadioReceiver.cs:
```diff
@@ -0,0 +1,63 @@
+
+	public void OnReceiveSignal(RadioSignal signal)
+	{
+		signalCallbacks.Invoke(signal);
```
Should there be a null check here or is this safe?

RedlineTriad commented on UnityProject/Assets/Scripts/UI/Hacking/GUI_HackingNode.cs:
```diff
@@ -0,0 +1,49 @@
+		parentHackingPanel = GetComponentInParent<GUI_Hacking>();
+	}
+
+	public void SetHackingNode(HackingNode node)
+	{
+		hackNode = node;
+		SetUpNodeData();
+	}
```
Add a setter to `public HackingNode HackNode => hackNode;` instead.

RedlineTriad commented on UnityProject/Assets/Scripts/UI/Hacking/GUI_HackingNode.cs:
```diff
@@ -0,0 +1,49 @@
+		SetUpNodeData();
+	}
+
+	public void SetUpNodeData()
+	{
+		//label.SetText(hackNode.HiddenLabel);
+	}
```
What is this needed for?

RedlineTriad commented on UnityProject/Assets/Scripts/UI/Hacking/GUI_HackingWire.cs:
```diff
@@ -0,0 +1,78 @@
+		parentHackingPanel = GetComponentInParent<GUI_Hacking>();
+	}
+
+	public void SetStartUINode(GUI_HackingNode startNode)
+	{
+		this.startNode = startNode;
+		RectTransform nodeRectTransform = startNode.GetComponent<RectTransform>();
+		RectTransform wireStartRectTransform = wireStart.GetComponent<RectTransform>();
+
+		wireStartRectTransform.sizeDelta = nodeRectTransform.sizeDelta * 0.7f;
+
+		wireStartRectTransform.position = nodeRectTransform.position;
+	}
```
Add a setter instead.

RedlineTriad commented on UnityProject/Assets/Scripts/UI/Hacking/GUI_HackingWire.cs:
```diff
@@ -0,0 +1,78 @@
+		wireStartRectTransform.position = nodeRectTransform.position;
+	}
+
+	public void SetEndUINode(GUI_HackingNode endNode)
+	{
+		this.endNode = endNode;
+		RectTransform nodeRectTransform = endNode.GetComponent<RectTransform>();
+		RectTransform wireEndRectTransform = wireEnd.GetComponent<RectTransform>();
+
+		wireEndRectTransform.sizeDelta = nodeRectTransform.sizeDelta * 0.7f;
+
+		wireEndRectTransform.position = nodeRectTransform.position;
+	}
```
Add a setter instead.

RedlineTriad commented on UnityProject/Assets/Scripts/UI/Hacking/GUI_HackingDevice.cs:
```diff
@@ -0,0 +1,41 @@
+		parentHackingPanel = GetComponentInParent<GUI_Hacking>();
+	}
+
+	public void SetHackingDevice(HackingDevice device)
+	{
+		this.device = device;
+		SetUpDeviceData();
+	}
```
Use a setter instead.

RedlineTriad commented on UnityProject/Assets/Scripts/Messages/Client/AddHackingConnection.cs:
```diff
@@ -0,0 +1,56 @@
+	public override void Process()
+	{
+		LoadMultipleObjects(new uint[] { Player, HackableObject });
+		int[] connectionToAdd = JsonConvert.DeserializeObject<int[]>(JsonData);
```
I am sure Mirror can serialize arrays without newtonsoft.

RedlineTriad reviewed (changes requested):
This is just a code-review from github, will do testing later too.

Fortune117 committed:
Small change to fix loop caused by inputs being wired together.

Alos fixed a null reference, due to the AcceptID node not having any registered methods to run.

Fortune117 committed:
Update GUI_Hacking.cs

Fortune117 committed:
Fixed small dual node issue. (acting as input/output)

Fortune117 committed:
Merge remote-tracking branch 'upstream/develop' into hacking

Fortune117 committed:
Merged with Upstream

DooblyNoobly committed:
Merge branch 'premergedev' into hacking
